File: runtimes/hermes/src/kars_runtime_hermes/plugin/mcp_bridge.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

"""Bounded MCP access through the existing loopback router client."""

from __future__ import annotations

import json
import re
import threading
from typing import Any

from . import router_client

_LOCK = threading.Lock()
_SLOTS = threading.BoundedSemaphore(8)
_SESSION: str | None = None
_INITIALIZED = False
_NEXT_ID = 1
_PROTOCOL = "2025-06-18"
_MAX_BYTES = 2 * 1024 * 1024


def _headers(server: str | None) -> dict[str, str]:
    headers = {"Accept": "application/json, text/event-stream", "MCP-Protocol-Version": _PROTOCOL}
    if _SESSION:
        headers["Mcp-Session-Id"] = _SESSION
    if server:
        headers["X-Kars-Mcp-Server"] = server
    return headers


def _request(method: str, params: dict[str, Any], server: str | None, notification: bool = False) -> tuple[Any, dict[str, Any]]:
    global _NEXT_ID, _INITIALIZED
    request_id = _NEXT_ID
    _NEXT_ID += 1
    body = {"jsonrpc": "2.0", "method": method, "params": params}
    if not notification:
        body["id"] = request_id
    if len(json.dumps(body).encode()) > _MAX_BYTES:
        raise RuntimeError("MCP request exceeds its byte limit")
    response = router_client.call("POST", "/mcp", json=body, headers=_headers(server))
    if response.status_code in (400, 404):
        _INITIALIZED = False
    if not 200 <= response.status_code < 300:
        raise RuntimeError(f"MCP HTTP {response.status_code}")
    if len(response.content) > _MAX_BYTES:
        raise RuntimeError("MCP response exceeds its byte limit")
    if notification:
        return response, {}
    try:
        value = response.json()
    except Exception:
        raise RuntimeError("MCP router response is not valid JSON") from None
    if not isinstance(value, dict) or value.get("jsonrpc") != "2.0" or value.get("id") != request_id \
            or value.get("error") or not isinstance(value.get("result"), dict):
        raise RuntimeError("MCP RPC failed or omitted its matching result")
    return response, value["result"]


def _initialize(server: str | None) -> None:
    global _SESSION, _INITIALIZED, _PROTOCOL
    _SESSION = None
    _PROTOCOL = "2025-06-18"
    response, result = _request("initialize", {
        "protocolVersion": _PROTOCOL, "capabilities": {},
        "clientInfo": {"name": "kars-runtime-hermes", "version": "0.1.0"},
    }, server)
    protocol = result.get("protocolVersion")
    if protocol not in ("2024-11-05", "2025-03-26", "2025-06-18", "2025-11-25"):
        raise RuntimeError("MCP negotiated an unsupported protocol")
    session = response.headers.get("mcp-session-id")
    if session is not None and (not session or len(session) > 1024 or not re.fullmatch(r"[\x21-\x7e]+", session)):
        raise RuntimeError("MCP session identifier is invalid")
    _SESSION, _PROTOCOL = session, protocol
    _request("notifications/initialized", {}, server, notification=True)
    _INITIALIZED = True


def _invoke(method: str, params: dict[str, Any], server: str | None = None) -> dict[str, Any]:
    if server is not None and (not isinstance(server, str) or len(server) > 253
                              or not re.fullmatch(r"[a-z0-9.-]+", server)):
        raise RuntimeError("Invalid MCP server scope")
    if not _SLOTS.acquire(blocking=False):
        raise RuntimeError("MCP bridge queue is full")
    locked = False
    try:
        locked = _LOCK.acquire(timeout=60)
        if not locked:
            raise RuntimeError("MCP bridge queue timed out")
        if not _INITIALIZED:
            _initialize(server)
        # No tools/call retry, including accepted JSON-RPC errors.
        _, result = _request(method, params, server)
        return result
    finally:
        if locked:
            _LOCK.release()
        _SLOTS.release()


def _failure(message: str) -> str:
    return json.dumps({"error": message, "isError": True})


def _list_tools(args: dict[str, Any], **_kwargs: Any) -> str:
    try:
        result = _invoke("tools/list", {}, args.get("server"))
        if not isinstance(result.get("tools"), list):
            return _failure("MCP catalog omitted tools")
        return json.dumps(result, separators=(",", ":"))
    except Exception:
        return _failure("MCP discovery failed; inspect the router's scoped status")


def _call_tool(args: dict[str, Any], **_kwargs: Any) -> str:
    name, arguments = args.get("name"), args.get("arguments", {})
    if not isinstance(name, str) or not name.strip():
        return _failure("name is required")
    if not isinstance(arguments, dict):
        return _failure("arguments must be an object")
    try:
        return json.dumps(_invoke("tools/call", {"name": name, "arguments": arguments}, args.get("server")), separators=(",", ":"))
    except Exception:
        return _failure("MCP invocation failed; calls are never automatically replayed")


def register(ctx: Any) -> None:
    for name, description, parameters, handler in (
        ("kars_mcp_list", "List qualified MCP tools mounted in this sandbox, optionally scoped to one server.",
         {"type": "object", "properties": {"server": {"type": "string"}}}, _list_tools),
        ("kars_mcp_call", "Call an exact MCP tool name returned by kars_mcp_list. No automatic call replay.",
         {"type": "object", "properties": {"name": {"type": "string"}, "arguments": {"type": "object"},
                                          "server": {"type": "string"}}, "required": ["name"]}, _call_tool),
    ):
        ctx.register_tool(name=name, toolset="kars_mcp",
                          schema={"name": name, "description": description, "parameters": parameters},
                          handler=handler, description=description)


def _reset_for_tests() -> None:
    global _SESSION, _INITIALIZED, _NEXT_ID, _PROTOCOL
    _SESSION, _INITIALIZED, _NEXT_ID, _PROTOCOL = None, False, 1, "2025-06-18"

File: runtimes/openclaw/src/core/agt-tools/mcp-bridge.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import * as http from "node:http";

const MAX_BYTES = 2 * 1024 * 1024;
const MAX_QUEUED = 8;
const PROTOCOL = "2025-06-18";
interface State { initialized: boolean; session?: string; protocol: string; nextId: number; queued: number; tail: Promise<void> }
const KEY = Symbol.for("kars-mcp-bridge");
const globals = globalThis as typeof globalThis & { [KEY]?: State };
const state = globals[KEY] ??= { initialized: false, protocol: PROTOCOL, nextId: 1, queued: 0, tail: Promise.resolve() };

interface Response { status: number; headers: http.IncomingHttpHeaders; body: unknown }
interface Result { content: Array<{ type: "text"; text: string }>; isError?: boolean }
interface Tool { name: string; label: string; description: string; parameters: Record<string, unknown>;
  execute(id: string, args?: Record<string, unknown>): Promise<Result> }

function request(payload: Record<string, unknown>, server?: string): Promise<Response> {
  const body = JSON.stringify(payload);
  if (Buffer.byteLength(body) > MAX_BYTES) return Promise.reject(new Error("MCP request exceeds its byte limit"));
  return new Promise((resolve, reject) => {
    const req = http.request({
      hostname: "127.0.0.1", port: 8443, path: "/mcp", method: "POST",
      headers: { accept: "application/json, text/event-stream", "content-type": "application/json",
        "content-length": Buffer.byteLength(body), "mcp-protocol-version": state.protocol,
        ...(state.session ? { "mcp-session-id": state.session } : {}),
        ...(server ? { "x-kars-mcp-server": server } : {}) },
    }, response => {
      const chunks: Buffer[] = [];
      let size = 0;
      response.on("data", (chunk: Buffer) => {
        size += chunk.length;
        if (size > MAX_BYTES) { req.destroy(new Error("MCP response exceeds its byte limit")); return; }
        chunks.push(chunk);
      });
      response.on("error", () => reject(new Error("MCP response transport failed")));
      response.on("end", () => {
        try {
          const bytes = Buffer.concat(chunks).toString("utf8");
          resolve({ status: response.statusCode ?? 500, headers: response.headers,
            body: bytes ? JSON.parse(bytes) : undefined });
        } catch { reject(new Error("MCP router response is not valid JSON")); }
      });
    });
    req.on("error", error => reject(new Error(error.message.startsWith("MCP ") ? error.message : "MCP transport failed")));
    req.setTimeout(30_000, () => req.destroy(new Error("MCP request timed out")));
    req.end(body);
  });
}

function rpc(response: Response, id: number): Record<string, unknown> {
  if (response.status < 200 || response.status >= 300) throw new Error(`MCP HTTP ${response.status}`);
  const body = response.body as { jsonrpc?: unknown; id?: unknown; error?: unknown; result?: unknown } | undefined;
  if (body?.jsonrpc !== "2.0" || body.id !== id || body.error
    || !body.result || typeof body.result !== "object" || Array.isArray(body.result)) {
    throw new Error("MCP RPC failed or omitted its matching result");
  }
  return body.result as Record<string, unknown>;
}

async function initialize(server?: string): Promise<void> {
  state.session = undefined;
  state.protocol = PROTOCOL;
  const id = state.nextId++;
  const response = await request({ jsonrpc: "2.0", id, method: "initialize",
    params: { protocolVersion: PROTOCOL, capabilities: {}, clientInfo: { name: "kars-runtime-openclaw", version: "0.1.0" } } }, server);
  const result = rpc(response, id);
  if (typeof result.protocolVersion !== "string"
    || !["2024-11-05", "2025-03-26", "2025-06-18", "2025-11-25"].includes(result.protocolVersion)) {
    throw new Error("MCP negotiated an unsupported protocol");
  }
  const session = response.headers["mcp-session-id"];
  if (session !== undefined && (typeof session !== "string" || !session.length
    || session.length > 1024 || /[^\x21-\x7e]/.test(session))) throw new Error("MCP session identifier is invalid");
  state.session = session;
  state.protocol = result.protocolVersion;
  const notification = await request({ jsonrpc: "2.0", method: "notifications/initialized", params: {} }, server);
  if (notification.status < 200 || notification.status >= 300) throw new Error(`MCP initialization HTTP ${notification.status}`);
  state.initialized = true;
}

export async function invokeMcp(method: "tools/list" | "tools/call", params: Record<string, unknown>, server?: string): Promise<Record<string, unknown>> {
  if (server !== undefined && (typeof server !== "string" || !server.length || server.length > 253 || !/^[a-z0-9.-]+$/.test(server))) {
    throw new Error("Invalid MCP server scope");
  }
  if (state.queued >= MAX_QUEUED) throw new Error("MCP bridge queue is full");
  state.queued++;
  const deadline = Date.now() + 60_000;
  const run = async () => {
    try {
      if (Date.now() >= deadline) throw new Error("MCP bridge queue timed out");
      if (!state.initialized) await initialize(server);
      const id = state.nextId++;
      const response = await request({ jsonrpc: "2.0", id, method, params }, server);
      // A failed call is never replayed. A later explicit caller may open a
      // new session after a definite transport-level session rejection.
      if (response.status === 404 || response.status === 400) state.initialized = false;
      return rpc(response, id);
    } finally { state.queued--; }
  };
  const result = state.tail.then(run, run);
  state.tail = result.then(() => undefined, () => undefined);
  return result;
}

function text(value: unknown, failed = false): Result {
  return { content: [{ type: "text", text: JSON.stringify(value) }],
    ...(failed || (value as { isError?: unknown } | undefined)?.isError === true ? { isError: true } : {}) };
}

export function registerMcpBridgeTools(api: { registerTool(tool: Tool): void }): void {
  api.registerTool({
    name: "kars_mcp_list", label: "Kars MCP Tool List",
    description: "List qualified MCP tools mounted in this sandbox. Optional server scope narrows the catalog.",
    parameters: { type: "object", properties: { server: { type: "string" } } },
    async execute(_id, args = {}) {
      try {
        const result = await invokeMcp("tools/list", {}, args.server as string | undefined);
        if (!Array.isArray(result.tools)) throw new Error("MCP catalog omitted tools");
        return text(result);
      } catch (error) { return text({ error: error instanceof Error ? error.message : "MCP discovery failed" }, true); }
    },
  });
  api.registerTool({
    name: "kars_mcp_call", label: "Kars MCP Tool Call",
    description: "Invoke the exact MCP tool name returned by kars_mcp_list. Calls are never automatically replayed.",
    parameters: { type: "object", properties: { name: { type: "string" }, arguments: { type: "object" }, server: { type: "string" } }, required: ["name"] },
    async execute(_id, args = {}) {
      if (typeof args.name !== "string" || !args.name.trim()) return text({ error: "name is required" }, true);
      const argumentsValue = args.arguments ?? {};
      if (typeof argumentsValue !== "object" || argumentsValue === null || Array.isArray(argumentsValue)) {
        return text({ error: "arguments must be an object" }, true);
      }
      try { return text(await invokeMcp("tools/call", { name: args.name, arguments: argumentsValue }, args.server as string | undefined)); }
      catch (error) { return text({ error: error instanceof Error ? error.message : "MCP invocation failed" }, true); }
    },
  });
}

export function resetMcpBridgeForTests(): void {
  state.initialized = false;
  state.session = undefined;
  state.protocol = PROTOCOL;
  state.nextId = 1;
  state.queued = 0;
  state.tail = Promise.resolve();
}
